File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/21-HashTable/LeetCode/LeetCode-0705.cpp
class MyHashSet {
    #define TABLE_SIZE 2011 //1013 //5011    // Some prime number
    vector<list<int>> T;

public:
    /** Initialize your data structure here. */
    MyHashSet() {T.resize(TABLE_SIZE);}

    void add(int key) {
        int index = key % TABLE_SIZE;
        for (auto no: T[index]){
            if (no == key) return;   // Duplicates not allowed
        } //end-for

        T[index].push_back(key);
    } //end-add

    void remove(int key) {
        int index = key % TABLE_SIZE;
        for (auto iter = T[index].begin(); iter != T[index].end(); iter++){
            if (*iter == key){
                T[index].erase(iter);
                return;
            } //end-if
        } //end-for
    } //end-remove

    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        int index = key % TABLE_SIZE;
        for (auto no: T[index]){
            if (no == key) return true;   // Exists
        } //end-for

        return false;
    } // end-contains
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/21-HashTable/LeetCode/LeetCode-0705.java
class MyHashSet {
    private int TABLE_SIZE = 2011; //1013 //5011    // Some prime number
    private List<Integer>[] T = new LinkedList[TABLE_SIZE];

    public MyHashSet() {}
    
    public void add(int key) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) T[index] = new LinkedList<>();
        else if (T[index].contains(key)) return; // Duplicates not allowed
        T[index].add(key);
    }
    
    public void remove(int key) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) return;
        for (int i=0; i<T[index].size(); i++){
            if (T[index].get(i) == key){
                T[index].remove(i);
                return;
            } // end-for
        } // end-for
    }
    
    public boolean contains(int key) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) return false;
        return T[index].contains(key);
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */